File: lab3/3.2/cs/ImageDownloader/ImageDownloader/ImageDownloader.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;

namespace ImageDownloader
{
    internal class ImageDownloader
    {

        private readonly HttpClient _client;
        private readonly IFileSystemHandler _fileSystem;

        private ImageResponse _imageResponse;

        public ImageDownloader(HttpClient client, IFileSystemHandler fileSystem)
        {
            _client = client;
            _fileSystem = fileSystem;
        }

        public async Task<ImageResponse> GetNewFile(CancellationToken token = default)
        {
            var response = await _client.GetAsync("https://www.google.com", token);
            if (response.IsSuccessStatusCode)
            {
                var stream = await response.Content.ReadAsStringAsync(token);
                _imageResponse = JsonSerializer.Deserialize<ImageResponse>(stream);
                return _imageResponse;
            }
            return JsonSerializer.Deserialize<ImageResponse>("");
        }

        public bool WriteImageToDisk(string path, string input)
        {
            return _fileSystem.WriteLine(path, input) && _fileSystem.Exists(path);
        }

        public ImageResponse GetLastImageResponse()
        {
            return _imageResponse;
        }

        public class ImageResponse
        {
            public string msg { get; set; }
            public string url { get; set; }

            public ImageResponse(string msg, string url)
            {
                this.msg = msg;
                this.url = url;
            }
        }

    }
}

File: lab3/3.2/python/ImageDownloader.py
import json


class ImageDownloader:
    class ImageResponse:
        def __init__(self, msg: str, url: str):
            self.msg = msg
            self.url = url

    def __init__(self, client, file_system):
        self._client = client
        self._file_system = file_system
        self._image_response = None

    async def get_new_file(self, token=None):
        response = await self._client.get("https://www.google.com", token=token)
        if response.status_code == 200:
            stream = await response.content.read()
            self._image_response = json.loads(stream, object_hook=decode_response)
            return self._image_response
        return ImageDownloader.ImageResponse("", "")

    def write_image_to_disk(self, path, input):
        return self._file_system.write_line(path, input) and self._file_system.exists(path)

    def get_last_image_response(self):
        return self._image_response


def decode_response(obj):
    if 'msg' in obj and 'url' in obj:
        return ImageDownloader.ImageResponse(obj['msg'], obj['url'])